send* RPCs in the wallet returns the "fee reason"

stackman27 opened the pull request:
Whenever a wallet funds a transaction, the fee reason is reported to the user only if the verbose is set to true. I added an extra parameter to `CreateTransaction` function in wallet.cpp. Then I implemented the fee reason return logic in `SendMoney`  in rpcwallet.cpp, followed by verbose parameter in `sendtoaddress` and `sendmany` functions. I also added a fee reason test case in walletbasic.py.

link to the issue: https://github.com/MarcoFalke/bitcoin-core/issues/22#issue-616251578

stackman27 committed:
[send] Make send RPCs return fee reason

gzhao408 commented:
Concept ACK!

achow101 commented on src/wallet/feebumper.cpp:
```diff
@@ -219,7 +219,8 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
     int change_pos_in_out = -1; // No requested location for change
     bilingual_str fail_reason;
-    if (!wallet.CreateTransaction(recipients, tx_new, fee_ret, change_pos_in_out, fail_reason, new_coin_control, false)) {
+    std::string feeReason;
```
nit: Our naming convention is to use `snake_case` rather than `camelCase`.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -401,11 +401,20 @@ UniValue SendMoney(CWallet* const pwallet, const CCoinControl &coin_control, std
+
+    if(verbose){
+        entry.pushKV("hex", tx->GetHash().GetHex());
+        entry.pushKV("Fee Reason", feeReason);  
```
I don't think we put spaces as the key in a json object as spaces tend to break things. Typically we do lowercase snake_case.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -430,6 +439,8 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
                     {"avoid_reuse", RPCArg::Type::BOOL, /* default */ "true", "(only available if avoid_reuse wallet flag is set) Avoid spending from dirty addresses; addresses are considered\n"
             "                             dirty if they have previously been used in a transaction."},
+                    {"verbose", RPCArg::Type::BOOL, /* default */ "false",
+                            "Whether to display the fee reason or not."},
```
It  would be preferable to have this text be generic instead of specific for the fee reason. Same for the other RPCs you've changed.

achow101 reviewed (commented):
In order for cli to work, you will need to also add the new arguments to `src/rpc/client.cpp`.

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -401,11 +401,20 @@ UniValue SendMoney(CWallet* const pwallet, const CCoinControl &coin_control, std
     bilingual_str error;
     CTransactionRef tx;
-    bool fCreated = pwallet->CreateTransaction(recipients, tx, nFeeRequired, nChangePosRet, error, coin_control, !pwallet->IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS));
+    UniValue entry(UniValue::VOBJ);
```
nit: I think this can be put in the (smaller) `if (verbose)` scope

maflcko reviewed (approved):
Concept ACK

stackman27 commented:
Added a new commit: added new arguments to `src/rpc/client.cpp` . Changed the case to snake_case.  Also generalized the verbose description and added entry object inside verbose if.

gzhao408 commented on test/functional/wallet_basic.py:
```diff
@@ -653,6 +653,16 @@ def run_test(self):
         assert_array_result(tx["details"], {"category": "receive"}, expected_receive_vout)
         assert_equal(tx[verbose_field], self.nodes[0].decoderawtransaction(tx["hex"]))
 
+        self.log.info("Testing Fee Reason")
```
nit: log message can be more informative, like "Test send* RPCs with verbose=True"

gzhao408 replied:
Also then comments are not _that_ necessary since it's pretty self-explanatory what you're doing

gzhao408 replied:
Style-wise, if you keep comments, it's best for them to be complete sentences

gzhao408 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -430,6 +439,8 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
                     {"avoid_reuse", RPCArg::Type::BOOL, /* default */ "true", "(only available if avoid_reuse wallet flag is set) Avoid spending from dirty addresses; addresses are considered\n"
             "                             dirty if they have previously been used in a transaction."},
+                    {"verbose", RPCArg::Type::BOOL, /* default */ "false",
+                            "Optional, the fee reason (only present when `verbose` is passed)"},
```
This is just a tad bit confusing, I'd recommend something like "If true, return the fee reason"

gzhao408 replied:
Er, something more generic would be "If true, return extra information about the transaction created."

gzhao408 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -831,6 +843,8 @@ static UniValue sendmany(const JSONRPCRequest& request)
                     {"estimate_mode", RPCArg::Type::STR, /* default */ "unset", std::string() + "The fee estimate mode, must be one of (case insensitive):\n"
             "       \"" + FeeModes("\"\n\"") + "\""},
+                    {"verbose", RPCArg::Type::BOOL, /* default */ "false",
+                            "Optional, the fee reason (only present when `verbose` is passed)"},
```
Same

glozow reviewed (commented):
utACK

Travis is failing on the linter though, most likely whitespace. You can run it yourself with `test/lint/lint-all.sh` (you might need to install a few packages). I left a bunch of nits that you can include with your changes, and then please [rebase and squash](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits).

fjahr commented:
Concept ACK

Please squash the fixup commits.

stackman27 commented:
Update: Fixed the verbose information and also the log info in wallet_basic.py, also squashed all the nit commits.

fjahr commented:
> Update: Fixed verbose information and also the log info in wallet_basic.py, also squashed all the nit commits.

Thank you! But I meant to squash the fix-ups into the original commits. Otherwise, it's harder to review because reviewers can not just go through the commits in order but need to jump back and forth to check if what they see in the first commit is not already changed in a later commit. So, for example `fee_reason` should be already used in the first commit.

fjahr commented on test/functional/wallet_basic.py:
```diff
@@ -653,6 +653,16 @@ def run_test(self):
+        assert_equal(str(txid_feeReason_one["Fee Reason"]), "Fallback fee")
+
+        #testing send_rpc_verbose sendtoaddress
+        txid_feeReason_two = self.nodes[2].sendmany(dummy = '', amounts = {address: 10}, verbose = True) 
```
Linter fails here on the trailing whitespace.

fjahr commented on test/functional/wallet_basic.py:
```diff
@@ -653,6 +653,16 @@ def run_test(self):
 
+        self.log.info("Testing Fee Reason")
+        #testing send_rpc_verbose sendtoaddress
+        address = self.nodes[0].getnewaddress("test") 
```
Also here trailing whitespace.

stackman27 commented:
Squashed nit commits to [test] send rpcs returns fee reason!

laanwj commented on src/wallet/rpcwallet.cpp:
```diff
@@ -400,12 +400,21 @@ UniValue SendMoney(CWallet* const pwallet, const CCoinControl &coin_control, std
     }
     pwallet->CommitTransaction(tx, std::move(map_value), {} /* orderForm */);
+
+    if(verbose){
```
please follow the style: `if (verbose){ {`

laanwj commented:
Small comment on the commit messages
```
* dbaf… [test] send rpcs returns fee reason (Sishir Giri) (pull/19501/head)
* 7298… [send] send rpcs returns fee reason (Sishir Giri)
```
I think, instead of using the same message with a different prefix, it would be clearer to make it:

- [test] Check that send RPCs return fee reason
- [rpc] Make send RPCs return fee reason

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -400,12 +400,21 @@ UniValue SendMoney(CWallet* const pwallet, const CCoinControl &coin_control, std
+
+    if(verbose){
+        UniValue entry(UniValue::VOBJ);
+        entry.pushKV("hex", tx->GetHash().GetHex());
```
`hex` usually implies the full raw transaction. This is just a txid, so this should instead be named `txid`.

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -430,6 +439,8 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
+                    {"verbose", RPCArg::Type::BOOL, /* default */ "false",
+                            "Optional, If true, return extra information about the transaction."},
                 },
                 RPCResult{
```
The `RPCResult` will need to be updated to include the `verbose` result too.

achow101 replied:
This isn't quite what I was looking for.

There are 2 possible results, so we need to have different RPC results for them. `getrawtransaction` has a similar `verbose` result. I would suggest looking at that to see how this should be done.

achow101 commented on test/functional/wallet_basic.py:
```diff
@@ -653,5 +653,12 @@ def run_test(self):
+        self.log.info("Test send* RPCs with verbose=True")
+        address = self.nodes[0].getnewaddress("test")
+        txid_feeReason_one = self.nodes[2].sendtoaddress(address = address, amount = 10, verbose = True)
+        assert_equal(str(txid_feeReason_one["fee_reason"]), "Fallback fee")
```
The result should already be a `str` so there's no need for `str(...)` here.

stackman27 commented:
Resolved all the comments! Fixed travis and squash the nit commits!

achow101 commented on src/wallet/rpcwallet.cpp:
```diff
@@ -400,21 +400,19 @@ UniValue SendMoney(CWallet* const pwallet, const CCoinControl &coin_control, std
         UniValue entry(UniValue::VOBJ);
-        entry.pushKV("hex", tx->GetHash().GetHex());
-        entry.pushKV("fee_reason", fee_reason);  
-        return entry;    
```
seems like these changes should be in the previous commit

luke-jr commented:
Thoughts on turning `sendmany` param 1/dummy into an options object? (And not including this functionality in `sendtoaddress`)

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2701,7 +2702,7 @@ OutputType CWallet::TransactionChangeType(const Optional<OutputType>& change_typ
 }
 
-bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransactionRef& tx, CAmount& nFeeRet, int& nChangePosInOut, bilingual_str& error, const CCoinControl& coin_control, bool sign)
+bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CTransactionRef& tx, CAmount& nFeeRet, int& nChangePosInOut, bilingual_str& error, const CCoinControl& coin_control, std::string& fee_reason, bool sign)
```
Seems like it would work better to use `FeeReason*` here (or perhaps `FeeCalculation*`)

stackman27 replied:
Are you recommending changing the pointer name from fee_reason to FeeReason?

fjahr replied:
He is referring to the enum `FeeReason` and the struct `FeeCalculation` to be used here instead of a string.

stackman27 replied:
How would i go about that because i'd have to change everything that calls `createTransaction` and also the UniValue entry string type? I'm just a bit confused because when I change its type to `FeeReason*', I get bombarded with errors

fjahr replied:
There is a bit more to it than just changing the type here in this line. The other places where you pass in a string now you will need to use the `FeeReason*` type as well. You also won't convert the reason to a string in this function but where you actually need it as a string. But most importantly: take your time to and try to understand these errors. Once you understand them they can guide you to the right solution.

stackman27 replied:
I changed the `str fee reason` to struct `FeeCalculation`

instagibbs commented on ~:
```diff
@@ -0,0 +1,29 @@
+pick c321fe670 [send] Make send RPCs return fee reason
```
uhhh, this looks like an accidental file addition? :)

stackman27 replied:
Fixed it

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -831,10 +845,16 @@ static UniValue sendmany(const JSONRPCRequest& request)
                     {"estimate_mode", RPCArg::Type::STR, /* default */ "unset", std::string() + "The fee estimate mode, must be one of (case insensitive):\n"
             "       \"" + FeeModes("\"\n\"") + "\""},
+                    {"verbose", RPCArg::Type::BOOL, /* default */ "false",
+                            "Optional, If true, return extra information about the transaction."},
```
`s/Optional, //` no need to say optional.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -430,9 +437,15 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
                 },
                 RPCResult{
-                    RPCResult::Type::STR_HEX, "txid", "The transaction id."
+                    RPCResult::Type::OBJ, "", "",
```
yay for object instead of string result, can add more things going forward.

MarcoFalke replied:
This will need to be two RPCResults (one for verbose=False and one for verbose=True). Refer to other places in the code base on how to do that.

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -430,9 +437,15 @@ static UniValue sendtoaddress(const JSONRPCRequest& request)
                     {"avoid_reuse", RPCArg::Type::BOOL, /* default */ "true", "(only available if avoid_reuse wallet flag is set) Avoid spending from dirty addresses; addresses are considered\n"
             "                             dirty if they have previously been used in a transaction."},
+                    {"verbose", RPCArg::Type::BOOL, /* default */ "false",
+                            "Optional, If true, return extra information about the transaction."},
```
`s/Optional, //` no need to say optional.

instagibbs reviewed (commented):
concept ACK

@luke-jr Previously I was told due to named args we don't have to stick everything in an object argument.

laanwj commented:
Travis:
```
* Checking consistency between dispatch tables and vRPCConvertParams

The subject line of commit hash c8992e89594a54edf283e4916f794475070b5114 is followed by a non-empty line. Subject lines should always be followed by a blank line.
```
(you need an empty line between the commit title and the body)

fjahr commented:
> Travis:
> 
> ```
> * Checking consistency between dispatch tables and vRPCConvertParams
> 
> The subject line of commit hash c8992e89594a54edf283e4916f794475070b5114 is followed by a non-empty line. Subject lines should always be followed by a blank line.
> ```
> 
> (you need an empty line between the commit title and the body)

It's not his fault, the new linter was buggy. The offending commit https://github.com/bitcoin/bitcoin/commit/c8992e89594a54edf283e4916f794475070b5114 is not part of this PR. Should be fixed with https://github.com/bitcoin/bitcoin/pull/19654. I restarted the job and it should pass that linter.

stackman27 commented:
Rebased and Changed `str feereason` to `FeeCalculation*`

stackman27 commented:
Rebased and fixed all the changes

MarcoFalke commented on src/wallet/wallet.h:
```diff
@@ -974,7 +974,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
      * @note passing nChangePosInOut as -1 will result in setting a random position
      */
-    bool CreateTransaction(const std::vector<CRecipient>& vecSend, CTransactionRef& tx, CAmount& nFeeRet, int& nChangePosInOut, bilingual_str& error, const CCoinControl& coin_control, bool sign = true);
+    bool CreateTransaction(const std::vector<CRecipient>& vecSend, CTransactionRef& tx, CAmount& nFeeRet, int& nChangePosInOut, bilingual_str& error, const CCoinControl& coin_control, FeeCalculation& feecalc_feereason, bool sign = true);
```
```suggestion
    bool CreateTransaction(const std::vector<CRecipient>& vecSend, CTransactionRef& tx, CAmount& nFeeRet, int& nChangePosInOut, bilingual_str& error, const CCoinControl& coin_control, FeeCalculation& fee_calc_out, bool sign = true);
```

Could rename to `fee_calc_out` to clarify that this is an out param?

stackman27 replied:
Also added two RPCResult in `sendtoaddress` and `sendmany`

maflcko reviewed (approved):
Looks good, but the feedback about the rpc documentation should be fixed

MarcoFalke commented on src/wallet/rpcwallet.cpp:
```diff
@@ -853,11 +871,22 @@ static RPCHelpMan sendmany()
+                        {
+                            {RPCResult::Type::STR_HEX, "txid", "The transaction id for the send. Only 1 transaction is created regardless of\n"
+                "the number of addresses."},
+                            {RPCResult::Type::BOOL, "fee reason", "The transaction fee reason."}
```
```suggestion
                            {RPCResult::Type::STR, "fee reason", "The transaction fee reason."}
```

stackman27 replied:
code adjusted based on the feedback. Removed `str` from [test] and changed rpc result return to `str`

maflcko reviewed (approved):
ACK. Locks good beside feedback and https://github.com/bitcoin/bitcoin/pull/19501/files#r459066407

stackman27 committed:
[test] Make sure send rpc returns fee reason

MarcoFalke commented:
ACK a3fe42c140008b7dbbef55b3d75c56d2c6343c43

stackman27 commented:
@luke-jr @instagibbs @laanwj @fjahr  could you please verify whether I've resolved all the problems that you've highlighted? Thank you

instagibbs commented on test/functional/wallet_basic.py:
```diff
@@ -661,5 +661,13 @@ def run_test(self):
+        txid_feeReason_one = self.nodes[2].sendtoaddress(address = address, amount = 10, verbose = True)
+        assert_equal(txid_feeReason_one["fee_reason"], "Fallback fee")
+        txid_feeReason_two = self.nodes[2].sendmany(dummy = '', amounts = {address: 10}, verbose = True)
+        assert_equal(txid_feeReason_two["fee_reason"], "Fallback fee")
```
maybe add a case where you explicitly pass in `verbose = False` to cover that case too :+1:

stackman27 replied:
Added `verbose = False` case in test file

instagibbs reviewed (commented):
ACK a3fe42c140008b7dbbef55b3d75c56d2c6343c43

meshcollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -497,8 +514,9 @@ static RPCHelpMan sendtoaddress()
 
     std::vector<CRecipient> recipients;
     ParseRecipients(address_amounts, subtractFeeFromAmount, recipients);
+    bool verbose = request.params[9].isNull() ? false: request.params[9].get_bool();
```
micro-nit: spacing between `false:`

stackman27 replied:
for some reason git is ignoring my whitespace changes. Any suggestions on how to fix that?

meshcollider commented on src/wallet/rpcwallet.cpp:
```diff
@@ -853,11 +871,22 @@ static RPCHelpMan sendmany()
+                },
+                {
+                    RPCResult{"if verbose is not set or set to false",
+                        RPCResult::Type::STR_HEX, "txid", "The transaction id for the send. Only  1 transaction is created regardless of\n"
```
micro-nit: double space between `Only  1`

meshcollider reviewed (commented):
utACK 69cf5d4eeb73f7d685e915fc17af64634d88a4a2

instagibbs commented:
ACK https://github.com/bitcoin/bitcoin/pull/19501/commits/69cf5d4eeb73f7d685e915fc17af64634d88a4a2

meshcollider commented:
@stackman27 if there are a couple of ACKs then it's better not to fix small nits and leave them in case there's a bigger change that needs to be done. Small spacing nits, etc. won't stop a merge :smile:

stackman27 commented:
> @stackman27 if there are a couple of ACKs then it's better not to fix small nits and leave them in case there's a bigger change that needs to be done. Small spacing nits, etc. won't stop a merge 😄

Okay for sure! Thank you

MarcoFalke commented:
Mind force pushing the previous reviewed version?

```
git push your_remote 69cf5d4:send_rpc_feereason -f
```

?

stackman27 commented:
> Mind force pushing the previous reviewed version?
> 
> ```
> git push your_remote 69cf5d4:send_rpc_feereason -f
> ```
> 
> ?

Yes just pushed!
